dotnet-sdk_7: 7.0.201 -> 7.0.202

mdarocha opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

https://github.com/dotnet/core/blob/main/release-notes/7.0/7.0.4/7.0.4.md

This also introduces an extension to `buildDotnetModule`, allowing to use the result of `combinePackages` to build packages. This in turn fixes `BeatSaberModManager`, so that it does not require regenerating dependencies after a dotnet update. Note that for for BeatSaberModManager to build, https://github.com/NixOS/nixpkgs/pull/221998 with the .net 6 update needs to be merged first.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

raphaelr commented:
The netcoredbg issue was resolved by #221438 earlier today. Otherwise LGTM.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1966

raphaelr commented on pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:
```diff
@@ -37,7 +39,7 @@ for package in *; do
     id=$(xq -r .package.metadata.id "$version/$package".nuspec)
 
-    if [[ -e "$excluded_source/$id.$version".nupkg ]]; then
+    if grep -q "$id.$version.nupkg" "$excluded_list"; then
```
```suggestion
    if grep -qxF "$id.$version.nupkg" "$excluded_list"; then
```
* -x: Matches whole line (without this the pattern `x.1.0.0.nupkg` matches the file `fox.1.0.0.nupkg`
* -F: Interprets the pattern as a fixed string instead of a regular expression. Otherwise the `.` in the pattern will match any character.

mdarocha commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -121,11 +121,15 @@ let
+  sdkDeps = lib.lists.flatten [ dotnet-sdk.packages ];
+
+  sdkSource = let
+    version = if dotnet-sdk ? version
+              then dotnet-sdk.version
+              else lib.concatStringsSep "-" dotnet-sdk.passthru.versions;
```
unfortunately since `combinePackages` uses `buildEnv`, we can't set `version` on it. Maybe after a large refactor of `combinePackages` :)

mdarocha committed:
BeatSaberModManager: regenerate deps, remove add-runtime-identifier.patch

mdarocha committed:
dotnet-sdk_7: 7.0.201 -> 7.0.202

mdarocha committed:
buildDotnetModule: add support for using combinePackages as dotnet-sdk
This allows packages that require several dotnet versions to build (like
BeatSaberModManager) to properly depend on the dotnet-sdk specific deps.
This in turns avoids having to regenerate the deps of those packages
after each dotnet-sdk update.

This also changes nuget-to-nix to accept a file with a list of
exclusions instead of a folder.

raphaelr commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -121,11 +121,13 @@ let
-    name = "dotnet-sdk-${dotnet-sdk.version}-source";
-    deps = [ sdkDeps ];
+  sdkSource = let
+    version = dotnet-sdk.version or lib.concatStringsSep "-" dotnet-sdk.versions;
```
```suggestion
    version = dotnet-sdk.version or (lib.concatStringsSep "-" dotnet-sdk.versions);
```
The parentheses are important: e.g. `nix-build -A ArchiSteamFarm`:

```
error: attempt to call something which is not a function but a string

       at /home/xx/.cache/nixpkgs-review/rev-e7cc9afbba6fb8d0657ed70ecabe309a34d92640/nixpkgs/pkgs/build-support/dotnet/build-dotnet-module/default.nix:127:15:

          126|   sdkSource = let
          127|     version = dotnet-sdk.version or (lib.concatStringsSep "-" dotnet-sdk.versions);
             |               ^
          128|   in mkNugetSource {
```

SuperSandro2000 commented:
@ofborg eval
